library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c

```diff
@@ -0,0 +1,78 @@
+/**
+ *  Copyright Notice:
+ *  Copyright 2021-2022 DMTF. All rights reserved.
+ *  License: BSD 3-Clause License. For full text see link: https://github.com/DMTF/libspdm/blob/main/LICENSE.md
+ **/
+
+#include "internal/libspdm_crypt_lib.h"
+
+#if LIBSPDM_FIPS_MODE
+
+static const uint8_t key[32] = {0};
+
+static const uint8_t msg[32] = {0};
+
+static const uint8_t hmac_sha256_answer[32] = {
+    0x33, 0xad, 0x0a, 0x1c, 0x60, 0x7e, 0xc0, 0x3b,
+    0x09, 0xe6, 0xcd, 0x98, 0x93, 0x68, 0x0c, 0xe2,
+    0x10, 0xad, 0xf3, 0x00, 0xaa, 0x1f, 0x26, 0x60,
+    0xe1, 0xb2, 0x2e, 0x10, 0xf1, 0x70, 0xf9, 0x2a
+};
+
+static const uint8_t hmac_sha512_answer[64] = {
+    0xba, 0xe4, 0x6c, 0xeb, 0xeb, 0xbb, 0x90, 0x40,
+    0x9a, 0xbc, 0x5a, 0xcf, 0x7a, 0xc2, 0x1f, 0xdb,
+    0x33, 0x9c, 0x01, 0xce, 0x15, 0x19, 0x2c, 0x52,
+    0xfb, 0x9e, 0x8a, 0xa1, 0x1a, 0x8d, 0xe9, 0xa4,
+    0xea, 0x15, 0xa0, 0x45, 0xf2, 0xbe, 0x24, 0x5f,
+    0xbb, 0x98, 0x91, 0x6a, 0x9a, 0xe8, 0x1b, 0x35,
+    0x3e, 0x33, 0xb9, 0xc4, 0x2a, 0x55, 0x38, 0x0c,
+    0x51, 0x58, 0x24, 0x1d, 0xae, 0xb3, 0xc6, 0xdd
+};
+
+/**
+ * HMAC KAT test: HMAC-SHA512 KAT and HMAC-SHA256 KAT.
+ * HMAC-SHA512 KAT covers SHA384, SHA512, HMAC-SHA384, and HMAC-SHA512.
+ * HMAC-SHA256 KAT covers SHA1, SHA224, SHA256, HMAC-SHA1, HMAC-SHA224, and HMAC-SHA256.
+ **/
+bool libspdm_fips_selftest_hmac(void)
+{
+    uint8_t hmac_256_result[32];
+    uint8_t hmac_512_result[64];
+    bool result;
+    result = true;
+
+#if LIBSPDM_SHA256_SUPPORT
+    result = libspdm_hmac_sha256_all(msg, sizeof(msg), key, sizeof(key), hmac_256_result);
+    if (!result) {
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "gen hmac_sha256 failed \n"));

```

Comment:
Which algorithm failed needs to be conveyed to the caller, and it should probably not return early. For example if SHA-256 fails but SHA-512 passes then Integrator can use SHA-512 but not SHA-256.

Reply:
I have added the `status` as the final result.